test/e2e/k8s_util.go

```diff
@@ -1032,55 +1032,116 @@ func (k *KubernetesUtils) waitForPodInNamespace(ns string, pod string) ([]string
-			return false
+// Encapsulate the data needed to perform a probe between pods
+type probeVector struct {
+	fromPod  Pod
+	toPod    Pod
+	port     int32
+	protocol utils.AntreaPolicyProtocol
+}
+
+// Populate the channel with all combinations of probes based on the required ports and protocols
+func (hsr *httpServerReadiness) buildProbeVectors(probes chan<- probeVector) {
+	for protocol, ports := range hsr.protocolPortPairs {
+		for _, fromPod := range hsr.pods {
+			for _, podTo := range hsr.pods {
+				for _, port := range ports {
+					probes <- probeVector{fromPod, podTo, port, protocol}
+				}
+			}
 		}
-		return true
 	}
+	close(probes)
+}
 
-	for i := 0; i < maxTries; i++ {
-		if serversAreReady() {
-			log.Infof("All HTTP servers are ready")
-			return nil
-		}
-		time.Sleep(defaultInterval)
+// Calculate the number of probes created across all port protocol permutations
+func (hsr *httpServerReadiness) numProbes() int {
+	probeCount := 0
+	podCount := len(hsr.pods)
+	podCountSquared := podCount * podCount
+	for protocol := range hsr.protocolPortPairs {
+		ports := hsr.protocolPortPairs[protocol]
+		probeCount += podCountSquared * len(ports)
 	}
-	return fmt.Errorf("after %d tries, HTTP servers are not ready", maxTries)
+	return probeCount
 }
 
-func (k *KubernetesUtils) validateOnePort(allPods []Pod, reachability *Reachability, port int32, protocol utils.AntreaPolicyProtocol) {
-	numProbes := len(allPods) * len(allPods)
-	resultsCh := make(chan *probeResult, numProbes)
-	// TODO: find better metrics, this is only for POC.
-	oneProbe := func(podFrom, podTo Pod, port int32) {
+// Spawn a fixed set of workers to complete probing of the servers
+func (hsr *httpServerReadiness) spawnProberPool(resultsCh chan *probeResult) {
+	probes := make(chan probeVector, hsr.numProbes())
+	hsr.buildProbeVectors(probes)
+
+	probe := func(vector probeVector) {
+		podFrom := vector.fromPod
+		podTo := vector.toPod
+		port := vector.port
+		protocol := vector.protocol
 		log.Tracef("Probing: %s -> %s", podFrom, podTo)
-		expectedResult := reachability.Expected.Get(podFrom.String(), podTo.String())
-		connectivity, err := k.Probe(podFrom.Namespace(), podFrom.PodName(), podTo.Namespace(), podTo.PodName(), port, protocol, nil, &expectedResult)
+		expectedResult := hsr.reachability.Expected.Get(podFrom.String(), podTo.String())
+		connectivity, err := hsr.Probe(podFrom.Namespace(), podFrom.PodName(), podTo.Namespace(), podTo.PodName(), port, protocol, hsr.remoteCluster, &expectedResult)
 		resultsCh <- &probeResult{podFrom, podTo, connectivity, err}
 	}
-	for _, pod1 := range allPods {
-		for _, pod2 := range allPods {
-			go oneProbe(pod1, pod2, port)
+
+	startProber := func() {
+		for vector := range probes {
+			probe(vector)
 		}
 	}
-	for i := 0; i < numProbes; i++ {
+
+	// Tested value as the upper limit for running locally with minimal impacts to CI speeds
+	proberRateLimit := 150
+	for range min(proberRateLimit, hsr.numProbes()) {
```

Comment:
this is not exactly the same as stopping the creation of probers when there is no more work, as you may still be creating a new prober while the input channel is already empty (in theory). However, it's still an improvement compared to the previous version and you are still achieving your goal of capping the number of concurrent probes, so this is good enough for me.